Introduce local Error type rather than using Geozero::GeozeroError

michaelkirk opened the pull request:
## Motivation


Currently we have a circular dependency between the geozero and flatgeobuf crates. Some related discussion in
https://github.com/flatgeobuf/flatgeobuf/issues/283

This starts to detangle us from geozero by introducing our own error type and using that for non-geozero code. It also allows us to have an error enum more suited for our use cases.

## What changed

1. Introduce a new `flatgeobuf::Error` type and use it for code that doesn't depend on Geozero.
2. Relocate code that depends on Geozero into separate `geozero_integration` modules. This code still returns GeozeroError.
3. For tests and benchs, which might return either `flatgeobuf::Error` or `geozero::GeozeroError` and mapping between them is verbose, use `Box<dyn std::error::Error>`. Alternatively, I'd also be OK with using `unwrap` everywhere in test code, but I was trying not to rock the boat too much.

michaelkirk commented:
The corresponding geozero PR is here: https://github.com/georust/geozero/pull/175

pka commented:
I quickly skimmed through the changes, and have general comments before going into detail
1. Error type: :+1: :+1: 
2. I'm not sure about that. I recently grepped for flatgeobuf in the geozero "core" crate and found so few hits, that I think removing flatgeobuf from the geozero crate is the way to #283. I'll add a comment there with more details.
3. No comments yet

michaelkirk commented:
Rebased, fixed an old lint, and realized the FallibleStreamingIterator need not return a GeozeroError.

> I'm not sure about that. I recently grepped for flatgeobuf in the geozero "core" crate and found so few hits, that I think removing flatgeobuf from the geozero crate is the way to https://github.com/flatgeobuf/flatgeobuf/issues/283. I'll add a comment there with more details.

I'll respond in that thread in hopes of centralizing the discourse. I'm hopeful that at least some of this PR is useful regardless of which direction we take on the "big picture" of resolving the circular dependency.

michaelkirk commented:
@pka - would you prefer that I revert the `mod geozero_integration` for now?

pka commented on src/rust/src/error.rs:
```diff
@@ -0,0 +1,45 @@
+
+#[derive(Debug)]
+pub enum Error {
+    Malformed(&'static str),
```
Are there many different error texts? @nyurik argued that we shouldn't use freetext messages for errors at all.

nyurik replied:
@pka I think `static str` is better than `String` (no allocation, much faster errors, esp when no error logging is needed), but I also think that a dedicated error type is better than `static str` because the user can make decisions based on the error type, and the compiler will ensure the stability of those errors (as oppose to doing brittle error string comparison).

michaelkirk replied:
I think freetext messages can be OK if you don't intend for the user to handle the specific texts differently.

That said, I think in this case Malformed was doing more work than it should. See b7e7f73

Are the unwrap's controversial? We're already asserting the same things that guarantee the invariant within the `generate_bounds` method so I don't think it introduces any crash in practice. I could leverage the typesystem to return some kind of "NonEmptyVec", but I'm not sure it's worth it.

pka commented on src/rust/src/file_reader.rs:
```diff
@@ -351,6 +301,117 @@ impl<R: Read + Seek> FallibleStreamingIterator for FeatureIter<R, Seekable> {
     }
 }
 
+mod geozero_integration {
```
I would prefer `geozero_api`

pka commented on src/rust/src/file_writer.rs:
```diff
@@ -365,129 +319,186 @@ impl<'a> FgbWriter<'a> {
-impl FeatureProcessor for FgbWriter<'_> {
-    fn feature_end(&mut self, _idx: u64) -> Result<()> {
-        self.write_feature()
+mod geozero_integration {
```
`geozero_api`

pka commented on src/rust/src/http_reader.rs:
```diff
@@ -199,39 +185,40 @@ impl AsyncFeatureIter {
-            cnt += 1;
+}
+
+mod geozero_integration {
```
same

pka commented on src/rust/src/packed_r_tree.rs:
```diff
@@ -886,6 +893,7 @@ fn tree_processing() -> Result<()> {
     let mut fout = BufWriter::new(tempfile()?);
-    tree.process_index(&mut GeoJsonWriter::new(&mut fout))?;
+    tree.process_index(&mut GeoJsonWriter::new(&mut fout))
+        .unwrap();
```
unwrap?

pka commented on src/rust/tests/read.rs:
```diff
@@ -328,7 +317,7 @@ fn reader_type() -> Result<()> {
         reader: R,
     ) -> Result<FeatureIter<R, reader_trait::Seekable>> {
-        FgbReader::open(reader)?.select_all()
+        Ok(FgbReader::open(reader)?.select_all()?)
```
This is the same as before, right? Is that better style?

nyurik replied:
`?` here might call an `Into` trait, whereas simply returning a `Result` might not work in case the error type is different

pka reviewed (changes requested):
Apart from a forgotten `unwrap`, everything is a an improvement over the current state, so fine fore merging. A few nitpicks we can discuss if you like.

nyurik commented:
I'm a bit confused - why are you using boxed errors?  In my understanding, this slower approach is mostly used in apps (i.e. `anyhow` crate does that internally), whereas most libs rely on `thiserror` style - where each error is statically typed

michaelkirk commented:
Firstly, note that the boxed errors are only used in test code, so I'm not overly concerned about performance there. 

This test code represents a place where the two dependencies (geozero vs fgb) haven't been detangled - where the code could be propagating either GeozeroError or flatgeobuf::Error.

nyurik commented:
ah, ok, yes, test code is fine i guess (would it perhaps make sense to simply rely on anyhow then to re-use existing pattern?  Definitely not a blocker though).  Also another possible solution might be to use this (untested):

```rust
#[derive(thiserror)]
enum MixedError {
  #[error("GeoZero error {0}"),
  GzError(#[from] GeozeroError),
  #[error("FlatGeoBuf error {0}"),
  FgbError(#[from] flatgeobuf::Error),
}
```
